### DISCUSSION TITLE
Challenges migrating from `--experimental-specifier-resolution=node` to custom loader `--loader`

### DISCUSSION THREAD

**Discussion by vscheuber** [nodejs/help#4736] [Q&A] [answered]
I hope there is an easy solution for my challenge, let me give a quick overview:

[Frodo CLI](https://github.com/rockcarver/frodo-cli) is a command line tool based on node. Users can run it as a global NPM package or download a packaged binary (we use [pkg](https://github.com/vercel/pkg)) for linux, windows, and macos. Frodo CLI is based on [Frodo Library](https://github.com/rockcarver/frodo-lib). Both lib and cli are typescript projects (almost fully converted from CJS) and lib is a hybrid CJS/ESM library.

It's been an adventure getting it all to work and one of the last pieces that finally got it all to function properly was the `--experimental-specifier-resolution=node`.

So one morning after an update to my laptop the NPM version of Frodo was broken and I found out that I had fallen victim to a change in node 19: https://github.com/nodejs/node/pull/44859 . This is the error I was getting:
```
temp % frodo -v
node:internal/errors:490
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/vscheuber/Projects/frodo-cli/esm/cli/admin/admin' imported from /Users/vscheuber/Projects/frodo-cli/esm/app.js
    at __node_internal_captureLargerStackTrace (node:internal/errors:490:5)
    at new NodeError (node:internal/errors:399:5)
    at finalizeResolution (node:internal/modules/esm/resolve:231:11)
    at moduleResolve (node:internal/modules/esm/resolve:850:10)
    at defaultResolve (node:internal/modules/esm/resolve:1058:11)
    at nextResolve (node:internal/modules/esm/loader:164:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:419:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
    at link (node:internal/modules/esm/module_job:76:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v19.4.0
```

Reading through the PR description and comments I thought: wow, this might be a perfect opportunity to future-proof the tool and support node 19.

Here is the challenge I'm seeking suggestions and ideas for:

I've been using this as the first line in the `app.ts`/`index.js` file in the Frodo CLI project:
```
#!/usr/bin/env -S node --no-warnings --enable-source-maps --experimental-specifier-resolution=node
```

To start the migration to a custom loader, I added `commonjs-extension-resolution-loader` to the CLI project:
```
frodo-cli % npm i commonjs-extension-resolution-loader
```

Next I modified `app.ts`/`index.js`in Frodo CLI project:
```
#!/usr/bin/env -S node --no-warnings --enable-source-maps --loader commonjs-extension-resolution-loader
```

If I now run the CLI from within the project directory, it works perfectly:
```
frodo-cli % frodo -v      
You seem to be running the NPM package
Installed versions:
cli: v0.20.1
lib: v0.18.1
node: v19.4.0
```

If, however, I run the CLI (which is installed as a global NPM package) from any other directory, node cannot find the custom loader:
```
temp % frodo -v
node:internal/errors:490
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'commonjs-extension-resolution-loader' imported from /Users/vscheuber/temp/
    at __node_internal_captureLargerStackTrace (node:internal/errors:490:5)
    at new NodeError (node:internal/errors:399:5)
    at packageResolve (node:internal/modules/esm/resolve:794:9)
    at moduleResolve (node:internal/modules/esm/resolve:843:20)
    at defaultResolve (node:internal/modules/esm/resolve:1058:11)
    at nextResolve (node:internal/modules/esm/loader:164:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:419:18)
    at ESMLoader.import (node:internal/modules/esm/loader:520:22)
    at loadModulesInIsolation (node:internal/process/esm_loader:63:34)
    at initializeLoader (node:internal/process/esm_loader:32:25)
    at loadESM (node:internal/process/esm_loader:72:11)
    at runMainESM (node:internal/modules/run_main:54:21)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:80:5)
    at node:internal/main/run_main_module:23:47 {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v19.4.0
```

How can I allow node to locate the custom loader when I launch from any directory outside the project folder?

I greatly appreciate any suggestions. Thank y'all!

**Comment by aduh95** [comment, marked as answer]
The easiest workaround would be to use `--loader frodo/loader` (or `frodo/dist/loader.mjs` or similar if you're not using exports map) which would point to a file that does `export * from 'commonjs-extension-resolution-loader';` (assuming `commonjs-extension-resolution-loader` is listed in the deps of your npm package). IIRC, `frodo` installed globally means that Node.js should be able to resolve `frodo/loader` from anywhere, but maybe that's wrong, I don't know, I never use globally installed packages.

Another way is to use a data: URL that contains the whole loader, but that's not very practical.

**Reply by vscheuber** [reply]
Thank you for the suggestion, @aduh95 ! Here is what I tried, heeding your advice:

Made sure Frodo CLI is installed as a global package:
```
temp % npm list -g
/Users/vscheuber/.nvm/versions/node/v19.4.0/lib
├── @rockcarver/frodo-cli@0.20.1 -> ./../../../../../Projects/frodo-cli
├── corepack@0.15.2
└── npm@9.2.0
```

I created a loader.js file at the root of `@rockcarver/frodo-cli` which exports `commonjs-extension-resolution-loader` then updated the loader configuration to:
```
#!/usr/bin/env node --no-warnings --enable-source-maps --loader=@rockcarver/frodo-cli/loader.js
```

With that configuration, node cannot find the loader no matter from where I run it. From within the project directory:
```
frodo-cli % frodo
node:internal/errors:490
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package '@rockcarver/frodo-cli' imported from /Users/vscheuber/Projects/frodo-cli/
    at __node_internal_captureLargerStackTrace (node:internal/errors:490:5)
    at new NodeError (node:internal/errors:399:5)
    at packageResolve (node:internal/modules/esm/resolve:794:9)
    at moduleResolve (node:internal/modules/esm/resolve:843:20)
    at defaultResolve (node:internal/modules/esm/resolve:1058:11)
    at nextResolve (node:internal/modules/esm/loader:164:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:419:18)
    at ESMLoader.import (node:internal/modules/esm/loader:520:22)
    at loadModulesInIsolation (node:internal/process/esm_loader:63:34)
    at initializeLoader (node:internal/process/esm_loader:32:25)
    at loadESM (node:internal/process/esm_loader:72:11)
    at runMainESM (node:internal/modules/run_main:54:21)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:80:5)
    at node:internal/main/run_main_module:23:47 {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v19.4.0
```

Or from anywhere else on the system:
```
temp % frodo
node:internal/errors:490
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package '@rockcarver/frodo-cli' imported from /Users/vscheuber/temp/
    at __node_internal_captureLargerStackTrace (node:internal/errors:490:5)
    at new NodeError (node:internal/errors:399:5)
    at packageResolve (node:internal/modules/esm/resolve:794:9)
    at moduleResolve (node:internal/modules/esm/resolve:843:20)
    at defaultResolve (node:internal/modules/esm/resolve:1058:11)
    at nextResolve (node:internal/modules/esm/loader:164:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:419:18)
    at ESMLoader.import (node:internal/modules/esm/loader:520:22)
    at loadModulesInIsolation (node:internal/process/esm_loader:63:34)
    at initializeLoader (node:internal/process/esm_loader:32:25)
    at loadESM (node:internal/process/esm_loader:72:11)
    at runMainESM (node:internal/modules/run_main:54:21)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:80:5)
    at node:internal/main/run_main_module:23:47 {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v19.4.0
```

I believe I am not understanding how node tries to find packages or files specified via the `--loader` option. I am able to have it resolve a local package but only when launching node from within the project directory. But I cannot make it resolve a global package no matter from any launch location.

Just for kicks I'm going to try the data URL and also trying to get some debug output to better understand how the loader-loading works and why it fails.

Thank you again!

**Reply by vscheuber** [reply]
Hi @aduh95 , wanted to give a quick update on the data uri approach:

I tried a number of different mime types and base64 vs not but no matter which format I try:
```
data:text/javascript;charset=utf-8;base64,aW1wb3J0e2lzQnVpbHRpbn1mcm9tIm5vZGU6bW9kdWxlIjtpbXBvcnR7ZGlybmFtZX1mcm9tIm5vZGU6cGF0aCI7aW1wb3J0e2N3ZH1mcm9tIm5vZGU6cHJvY2VzcyI7aW1wb3J0e2ZpbGVVUkxUb1BhdGgscGF0aFRvRmlsZVVSTH1mcm9tIm5vZGU6dXJsIjtpbXBvcnR7cHJvbWlzaWZ5fWZyb20ibm9kZTp1dGlsIjtpbXBvcnQgcmVzb2x2ZUNhbGxiYWNrIGZyb20icmVzb2x2ZS9hc3luYy5qcyI7Y29uc3QgcmVzb2x2ZUFzeW5jPXByb21pc2lmeShyZXNvbHZlQ2FsbGJhY2spLGJhc2VVUkw9cGF0aFRvRmlsZVVSTChjd2QoKSsiLyIpLmhyZWY7ZXhwb3J0IGFzeW5jIGZ1bmN0aW9uIHJlc29sdmUobyxlLHIpe2NvbnN0e3BhcmVudFVSTDp0PWJhc2VVUkx9PWU7aWYoaXNCdWlsdGluKG8pKXJldHVybiByKG8sZSk7by5zdGFydHNXaXRoKCJmaWxlOi8vIikmJihvPWZpbGVVUkxUb1BhdGgobykpO2NvbnN0IGk9ZmlsZVVSTFRvUGF0aCh0KTtsZXQgczt0cnl7Y29uc3QgZT1hd2FpdCByZXNvbHZlQXN5bmMobyx7YmFzZWRpcjpkaXJuYW1lKGkpLGV4dGVuc2lvbnM6WyIuanMiLCIuanNvbiIsIi5ub2RlIiwiLm1qcyJdfSk7cz1wYXRoVG9GaWxlVVJMKGUpLmhyZWZ9Y2F0Y2gobyl7dGhyb3ciTU9EVUxFX05PVF9GT1VORCI9PT1vLmNvZGUmJihvLmNvZGU9IkVSUl9NT0RVTEVfTk9UX0ZPVU5EIiksb31yZXR1cm4gcihzLGUpfQ==
```

I tried different mime types (`text/javascript`, `text/plain`, `application/x-javascript`) and w/ or w/o quotes.

the shell always chokes on the data uri:
```
frodo-cli % frodo
zsh: /usr/local/bin/frodo: bad interpreter: /usr/bin/env -S node --no-warnings --enable-source-maps --loader=data:text/plain;charset=utf-8;base64,aW1wb3J0IHsgaXNCdWlsdGluIH0: exec format error
frodo-cli % frodo
zsh: /usr/local/bin/frodo: bad interpreter: /usr/bin/env -S node --no-warnings --enable-source-maps --loader="data:text/javascript;charset=utf-8;base64,aW1wb3J0IHsgaXNCdWlsd: exec format error
frodo-cli % frodo
zsh: /usr/local/bin/frodo: bad interpreter: /usr/bin/env -S node --no-warnings --enable-source-maps --loader=data:text/plain;charset=utf-8,import%20%7B%20isBuiltin%20%7D%20f: exec format error
frodo-cli % frodo
zsh: /usr/local/bin/frodo: bad interpreter: /usr/bin/env -S node --no-warnings --enable-source-maps --loader=data:application/x-javascript;base64,aW1wb3J0IHsgaXNCdWlsdGlu: exec format error
```

I couldn't find a way to have the shell let node interpret the string.

Not hopeful the data: uri approach will work but if you had a particular way in mind, please let me know.

Maybe the `--loader` option is not (yet) a full equivalent of the pre-node19 `--experimental-specifier-resolution=node` option?

**Reply by aduh95** [reply]
The correct MIME would be `text/javascript`, base64 is not necessary if you escape the URI special chars.

My guess regarding the `bad interpreter` error message means that ZSH is not able to parse the hashbang, so changing the MIME won't help you. I guess there's some char limit to how long a hashbang can be.



> I believe I am not understanding how node tries to find packages or files specified via the `--loader` option.

As [said in the docs](https://nodejs.org/api/cli.html#--experimental-loadermodule), the string can be anything accepted by `import()`, and then it uses the same algorithm as `import` which is also [described in the docs](https://nodejs.org/api/esm.html#resolver-algorithm-specification). The TL;DR of it for your use case is that it will search all `node_modules` folders starting from the cwd up until the root directory and gives up if it hasn't found anything.
According to npm docs, globally installed folders are located at `/usr/local/lib/node_modules` which could explain why you can't find them.

> Maybe the `--loader` option is not (yet) a full equivalent of the pre-node19 `--experimental-specifier-resolution=node` option?

It's very much not an equivalent, it provides all the tools to make Node.js behave the same as with the old flag, and more.

> Not hopeful the data: uri approach will work but if you had a particular way in mind, please let me know.

Since it looks like ZSH is the limiting factor, you can work around that by spawning another `node` process skipping the shell:

```js
#!/usr/bin/env node
'use strict';

const { spawn } = require('node:child_process');
const  { readFileSync } = require('node:fs');

const loaderContent = readFileSync(require.resolve('commonjs-extension-resolution-loader'));

spawn(process.execPath, [
  '--no-warnings', '--enable-source-maps',
  '--experimental-loader', `data:text/javascript;base64,${loaderContent.toString('base64')}`,
  require.resolve('frodo/actualBin'),
], { stdio: 'inherit', shell: false });
```

**Reply by vscheuber** [reply]
@aduh95 , let me just say: BRILLIANT!

I modified your sample to work as an ESM and that part seems to work great!:
```javascript
#!/usr/bin/env node
'use strict';

import { spawn } from 'node:child_process';
import { readFileSync } from 'node:fs';
import { createRequire } from 'module';

const require = createRequire(import.meta.url);
const loaderContent = readFileSync(
  require.resolve('commonjs-extension-resolution-loader')
);

spawn(
  process.execPath,
  [
    '--no-warnings',
    '--enable-source-maps',
    '--experimental-loader',
    `data:text/javascript;base64,${loaderContent.toString('base64')}`,
    './index.js',
  ],
  { stdio: 'inherit', shell: false }
);
```
Running frodo with that code in place, now generates another error, which looks like https://github.com/nodejs/node/issues/38790#issuecomment-878580942 but the proposed solution doesn't work for me. It's as if spawning node leads to a different behavior vs calling it the way I used to:
```
frodo-cli % frodo
node:internal/errors:490
    ErrorCaptureStackTrace(err);
    ^

TypeError [ERR_INVALID_URL]: Invalid URL
    at __node_internal_captureLargerStackTrace (node:internal/errors:490:5)
    at new NodeError (node:internal/errors:399:5)
    at URL.onParseError (node:internal/url:565:9)
    at new URL (node:internal/url:645:5)
    at getPackageScopeConfig (node:internal/modules/esm/package_config:106:24)
    at packageResolve (node:internal/modules/esm/resolve:748:25)
    at moduleResolve (node:internal/modules/esm/resolve:843:20)
    at defaultResolve (node:internal/modules/esm/resolve:1058:11)
    at nextResolve (node:internal/modules/esm/loader:164:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:419:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
    at link (node:internal/modules/esm/module_job:76:36)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  input: './package.json',
  code: 'ERR_INVALID_URL'
}

Node.js v19.4.0
```

I'll keep playing around with this some more and feel that a solution is neigh.

I can't thank you enough for all the help you have already provided!

**Reply by aduh95** [reply]
I'm glad it was helpful! Two notes:

- ESM is always in strict mode, you don't need to specify `"use strict"`.
- You should use `require.resolve` to get the absolute path of `./index.js', using a relative path for something that's supposed to work when spawn from anywhere seems dangerous.

The error you are getting is interesting, it should not be possible to trigger it, so I guess you stumbled upon an actual bug. If you were able to produce a minimal repro and open an issue for this, that would be awesome.

**Reply by vscheuber** [reply]
Oh! Yes, I guess I copied the code while in the midst of experimenting how to overcome that `ERR_INVALID_URL` error. I am using the `require.resolve` like your sample suggested. Thanks for the tip on `use strict`: removed.

**Reply by vscheuber** [reply]
@aduh95 I figured it out. Your idea of launching one instance of node and then spawning another made it click and so here comes the version that seems to work for frodo:

To resolve the `ERR_INVALID_URL` error, I had to drop the data uri part. With the spawning, my working directory problem (which really started this whole thread) was resolved. So the spawned node instance is now able to resolve the `../loader.js` module, which I included in the Frodo project directory, and thus I no longer need to worry about using data uris or playing other tricks around that.

I did hit another issue, which was that the child process did not inherit the CLI arguments passed to the parent process. I did solve that and made it work:

```javascript
#!/usr/bin/env node
import { spawn } from 'node:child_process';
import { createRequire } from 'module';

const require = createRequire(import.meta.url);

const launchArgs = [
  '--no-warnings',
  '--enable-source-maps',
  '--experimental-loader',
  require.resolve('../loader.js'),
  require.resolve('./index.js'),
];
const frodoArgs = process.argv.slice(2);

spawn(process.execPath, [...launchArgs, ...frodoArgs], {
  stdio: 'inherit',
  shell: false,
});
```
I'm sure this can be optimized some more but it appears to be working nicely. My handling of the CLI arguments (`frodoArgs`) feels a bit *hacky* but works.

Now my challenge will be to:
- make sure this works with node 14-19
- make sure this works with pkg, which creates the frodo binaries

Super grateful for your help, @aduh95 . Thank you!

**Reply by aduh95** [reply]
A small nit: `require.resolve` returns a path, and `--experimental-loader` expects a URL (using a path as a URL may works by accident because Unix paths look a lot like URLs as long as it doesn't contain any funky characters). The correct sample should be:

```js
#!/usr/bin/env node
import { spawn } from 'node:child_process';
import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);

const launchArgs = [
  '--no-warnings',
  '--enable-source-maps',
  '--experimental-loader',
  new URL('../loader.js', import.meta.url).href,
  require.resolve('./index.js'),
];
const frodoArgs = process.argv.slice(2);

spawn(process.execPath, [...launchArgs, ...frodoArgs], {
  stdio: 'inherit',
  shell: false,
});
```

Or

```js
#!/usr/bin/env node
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const launchArgs = [
  '--no-warnings',
  '--enable-source-maps',
  '--experimental-loader',
  new URL('../loader.js', import.meta.url).href,
  fileURLToPath(new URL('./index.js', import.meta.url)),
];
const frodoArgs = process.argv.slice(2);

spawn(process.execPath, [...launchArgs, ...frodoArgs], {
  stdio: 'inherit',
  shell: false,
});
```

**Reply by vscheuber** [reply]
Fantastic! Thank you, @aduh95 .

For completeness:

After implementing this launcher strategy I got everything to work great on the surface but later realized that my CLI tool was always returning exit code 0 no matter what. To overcome that, I modified the code slightly as follows:

``` javascript
#!/usr/bin/env node
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const launchArgs = [
  '--no-warnings',
  '--enable-source-maps',
  '--experimental-loader',
  new URL('./loader.js', import.meta.url).href,
  fileURLToPath(new URL('./app.js', import.meta.url)),
];
const frodoArgs = process.argv.slice(2);

const frodo = spawn(process.execPath, [...launchArgs, ...frodoArgs], {
  stdio: 'inherit',
  shell: false,
});

frodo.on('exit', (code) => {
  process.exitCode = code;
});
```

Now the exit code is transferred from the child back to the parent and all is well.
